# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.

In [61]:
! git clone https://www.github.com/DS3001/wrangling

Cloning into 'wrangling'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 36 (delta 13), reused 5 (delta 5), pack-reused 21
Receiving objects: 100% (36/36), 2.39 MiB | 16.43 MiB/s, done.
Resolving deltas: 100% (15/15), done.


**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1. The paper seems to be about why datasets should be tidied and the structure to approach doing so. In addition, the paper will introduce specific tools to help with the tidying process.
2. The tidy data standard is meant to accomplish a more efficient way of cleaning data by standardizing the process. This will include a way to look through and analyze the data so that the process in creating the data analysis tools will be easier.
3. The first sentence is meant to portray the difference between tidy and messy datasets--in a way, it's expanding upon the "tidy data standard." The first sentence is explaining that tidy data are similar because they've been cleaned in a "standardized" way that makes data analysis easier. Messy dataset is messy in its own way because we can never predict which areas of the dataset will need to be cleaned. The second sentence is stating that what we see in the labels of each column for datasets may not actually fully portray its signifance within the dataset. The examples they give are that labels could be seen as an observation or variable, but within the actual data set, may not be what it seems and actually has a different "definition". This is just showing the importance of understanding the context of the dataset in order to fully gauge the meaning of the variables and observations.
4. Wickham defines values as either quantitative (numbers) or qualitative (words/strings) data that can be sorted into variables or observations. Variables will describe the same trait regarding different subjects, so the example provided by Wickham is height, temperature, or duration. Observations, on the other hand, describes a single subject but the multiple traits that describe it. An example would be the observations that could be made regarding a person.
5. Tidy data is defined as a way "of mapping the meaning of a dataset to its structure" which is defining the way in which the data is formatted in order to be analyzed. Tidy data, specifically, is defined so that a variable will form a column while an observation will form a row. In addition, each type of observational unit will form a table. There is one type of tidy data since it standardizes a way of formatting the dataset so that it can be more efficiently utilized by a person or language.
6. The 5 most common problems with mess datasets are that column headers are values, not variables names, multiple variables being stored in one column, variables stored in both rows and columns, multiple types of observational units in the same table, and a single observational unit is stored in multiple tables. For table 4, it's messy because both rows and columns are composed of variables and the column headers don't have a name but have values instead. In order to fix this, we need to "melt" the data which means to switch the columns to rows. This means we can fix the problem of having both rows and columns as variables and having column headers that are values.
7. Table 11 is messy for a few reason. The first is that there is a lot of empty space throughout which were ommitted in Table 12. The melting part comes from creating column headers of id, year, month, and element. From there, tmax and tmin needed to be roated in order for them to be column headers rather than within the "element" column.
Overall, table 12 is tidy and "molten" because it's the cleaned version of table 11.
8. The chicken-and-egg problem relates to the paradoxical situation where the structure of the data and the tools that are created to be most efficient with the structure are interdepedent. Wickham hopes that in the future, we are able to use methodologies from other fields, such as user-testing, ethnography, and talk-aloud protocols, to improve the cognitive side and efficacy of tools for data analysis.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_NYC.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

1. I began by first reformatting all of the values under Price so that none of the numbers had commas in them. This was done with the replace function. From there, after getting the description of values, I saw that the type was an object which meant we wanted to coerce the data to be numeric, or a float, which was utilized with the code shown to us in class, specifically by using pd.to_numeric function. Even though I checked for missing values by utilizing hte imputation strategy that was given to us in class, the dataset didn't seem to have any missing values under Price. Just to check that, I used the Excel sheet and sorted the price column to check if there were any missing values, which there wasn't.
2. For the categorical variable, after using the unique function to check what types of variable names were under the type column, I saw that there were a few variables that could be consolidated into one. Specifically, boat, boatomg, and boating could all fit under watercraft. At first I was going to keep it as boating and watercraft, but felt that watercrafts was an umbrella term for boating. In addition, the actual raw data had descriptions of both watercrafts and boating accidents that had similar descriptions related to boats, so I combined them under one name. After that, unverified and under investigation both seemed to fit under unconfirmed, so I lumped them together. After that, I noticed that there was also "nan" which meant there was missing data, so I created a dummy value called "Unknown" that describes the missing data.
3.

In [3]:
! git clone https://github.com/yango4014/assignment2

Cloning into 'assignment2'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 20 (delta 3), reused 2 (delta 2), pack-reused 13
Receiving objects: 100% (20/20), 5.17 MiB | 18.89 MiB/s, done.
Resolving deltas: 100% (3/3), done.


In [27]:
import pandas as pd
df=pd.read_csv('assignment2/data/airbnb_NYC.csv', encoding="latin1")
var = 'Price'
df['Price']=df['Price'].str.replace(",", "") #reformat the numbers that have commas in them
#print(df.loc[263, "Price"])
print('Before coercion: \n', df[var].describe(),'\n')
df[var] = pd.to_numeric(df[var], errors='coerce') # Coerce the variable to numeric
df[var+'_nan'] = df[var].isnull() # Equals 1 if missing, 0 if non-null
print('After coercion: \n', df[var].describe(),'\n') # Describe the numeric variable
print('Total Missings: \n', sum(df[var+'_nan']),'\n') # How many missing values are there?


Before coercion: 
 count     30478
unique      511
top         150
freq       1481
Name: Price, dtype: object 

After coercion: 
 count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64 

Total Missings: 
 0 



In [58]:
import pandas as pd
import numpy as np

df=pd.read_csv('assignment2/data/sharks.csv', encoding="latin1")
var = 'Type'
df[var] = df[var].replace('Boatomg','Watercraft')
df[var] = df[var].replace('Boat','Watercraft')
df[var] = df[var].replace('Boating','Watercraft')
df[var] = df[var].replace('Unverified','Unconfirmed')
df[var] = df[var].replace('Under investigation','Unconfirmed')
df[var] = df[var].replace(np.nan,'Unknown')
#df[var] = df[var].fillna('Unknown', inplace=True)
print(df[var].unique(), '\n') # A Categorical Example
print(df[var].value_counts(), '\n')

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Invalid' 'Sea Disaster' 'Unknown'] 

Unprovoked      4716
Provoked         593
Invalid          552
Watercraft       344
Sea Disaster     239
Questionable      10
Unknown            5
Unconfirmed        3
Name: Type, dtype: int64 



<ipython-input-58-7396e0b21d8d>:4: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('assignment2/data/sharks.csv', encoding="latin1")


In [65]:
import pandas as pd
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url)
var = 'WhetherDefendantWasReleasedPretrial'
print(df[var].unique(), '\n')


KeyboardInterrupt: ignored

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_NYC.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.